pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -18,139 +18,254 @@
+        Object key = null;
+        Object value = null;
+        Schema<?> keySchema = null;
+        Schema<?> valueSchema = null;
+
+        if (record.getSchema() instanceof KeyValueSchema) {
+            KeyValueSchema<GenericObject,GenericObject> keyValueSchema = (KeyValueSchema) record.getSchema();
+            keySchema = keyValueSchema.getKeySchema();
+            valueSchema = keyValueSchema.getValueSchema();
+            if (KeyValueEncodingType.SEPARATED.equals(keyValueSchema.getKeyValueEncodingType()) && record.getKey().isPresent()) {
+                key = keySchema.decode(record.getKey().get().getBytes(StandardCharsets.UTF_8));
+            }
+            if (record.getValue() != null) {
+                value = ((KeyValue) record.getValue().getNativeObject()).getValue();
+                if (KeyValueEncodingType.INLINE.equals(keyValueSchema.getKeyValueEncodingType())) {
+                    key = ((KeyValue) record.getValue().getNativeObject()).getKey();
+                }
+            }
+        } else {
+            key = record.getKey().orElse(null);
+            valueSchema = record.getSchema();
+            value = record.getValue() == null ? null : record.getValue().getNativeObject();
+        }
 
-        if (!exists) {
-            CreateIndexRequest cireq = new CreateIndexRequest(elasticSearchConfig.getIndexName());
+        String id = key.toString();
+        if (keySchema != null) {
+            id = stringifyKey(keySchema, key);
+        }
 
-            cireq.settings(Settings.builder()
-               .put("index.number_of_shards", elasticSearchConfig.getIndexNumberOfShards())
-               .put("index.number_of_replicas", elasticSearchConfig.getIndexNumberOfReplicas()));
+        String doc = null;
+        if (value != null) {
+            if (valueSchema != null) {
+                doc = stringifyValue(valueSchema, value);
+            } else {
+                doc = value.toString();
+            }
+        }
 
-            CreateIndexResponse ciresp = getClient().indices().create(cireq, RequestOptions.DEFAULT);
-            if (!ciresp.isAcknowledged() || !ciresp.isShardsAcknowledged()) {
-                throw new RuntimeException("Unable to create index.");
+        if (elasticSearchConfig.isKeyIgnore()) {
+            if (doc == null || Strings.isNullOrEmpty(elasticSearchConfig.getPrimaryFields())) {
+                id = Hex.encodeHexString(record.getMessage().get().getMessageId().toByteArray());
+            } else {
+                try {
+                    // extract the PK from the JSON document
+                    JsonNode jsonNode = objectMapper.readTree(doc);
+                    List<String> pkFields = Arrays.asList(elasticSearchConfig.getPrimaryFields().split(","));
+                    id = stringifyKey(jsonNode, pkFields);
+                } catch (JsonProcessingException e) {
+                    log.error("Failed to read JSON", e);
```

Comment:
isn't this a hard error ?
if we are not able to process the key we may write garbage to ES